dialer.go

```diff
@@ -0,0 +1,99 @@
+// Copyright 2020 Google LLC
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+
+//     https://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cloudsqlconn
+
+import (
+	"context"
+	"crypto/rand"
+	"crypto/rsa"
+	"fmt"
+	"net"
+	"sync"
+
+	"cloud.google.com/cloudsqlconn/internal/cloudsql"
+	apiopt "google.golang.org/api/option"
+	sqladmin "google.golang.org/api/sqladmin/v1beta4"
+)
+
+type dialerConfig struct {
+	sqladminOpts []apiopt.ClientOption
+}
+
+// A Dialer is used to create connections to Cloud SQL instances.
+type Dialer struct {
+	lock      sync.RWMutex
+	instances map[string]*cloudsql.Instance
+	key       *rsa.PrivateKey
+
+	sqladmin *sqladmin.Service
+}
+
+// NewDialer creates a new Dialer.
+func NewDialer(ctx context.Context, opts ...DialerOption) (*Dialer, error) {
+	// TODO: Add shared / async key generation
+	key, err := rsa.GenerateKey(rand.Reader, 2048)
+	if err != nil {
+		return nil, fmt.Errorf("failed to generate rsa keys: %v", err)
+	}
+
+	var cfg dialerConfig
```

Comment:
Nit: I tend to see the `cfg := &dialerConfig{}` pattern more often when `cfg` is used as a pointer afterwords.